server/src/main/java/org/apache/iotdb/db/mpp/plan/analyze/PartitionCache.java

```diff
@@ -0,0 +1,710 @@
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+public class PartitionCache {
+  private static final Logger logger = LoggerFactory.getLogger(PartitionCache.class);
+  private static final IoTDBConfig config = IoTDBDescriptor.getInstance().getConfig();
+  private static final List<String> ROOT_PATH = Arrays.asList("root", "**");
+  private static final String STORAGE_GROUP_CACHE_NAME = "StorageGroup";
+  private static final String SCHEMA_PARTITION_CACHE_NAME = "SchemaPartition";
+  private static final String DATA_PARTITION_CACHE_NAME = "DataPartition";
+
+  /** calculate slotId by device */
+  private final String seriesSlotExecutorName = config.getSeriesPartitionExecutorClass();
+
+  private final int seriesPartitionSlotNum = config.getSeriesPartitionSlotNum();
+  private final SeriesPartitionExecutor partitionExecutor;
+
+  /** the size of partitionCache */
+  private final int cacheSize = config.getPartitionCacheSize();
+  /** the cache of storage group */
+  private final Set<String> storageGroupCache = Collections.synchronizedSet(new HashSet<>());
+  /** storage -> schemaPartitionTable */
+  private final Cache<String, SchemaPartitionTable> schemaPartitionCache;
+  /** storage -> dataPartitionTable */
+  private final Cache<String, DataPartitionTable> dataPartitionCache;
+
+  /** the latest time when groupIdToReplicaSetMap updated. */
+  private final AtomicLong latestUpdateTime = new AtomicLong(0);
+  /** TConsensusGroupId -> TRegionReplicaSet */
+  private final Map<TConsensusGroupId, TRegionReplicaSet> groupIdToReplicaSetMap =
+      new ConcurrentHashMap<>();
+
+  /** The lock of cache */
+  private final ReentrantReadWriteLock storageGroupCacheLock = new ReentrantReadWriteLock();
+
+  private final ReentrantReadWriteLock schemaPartitionCacheLock = new ReentrantReadWriteLock();
+  private final ReentrantReadWriteLock dataPartitionCacheLock = new ReentrantReadWriteLock();
+
+  private final IClientManager<PartitionRegionId, ConfigNodeClient> configNodeClientManager =
+      new IClientManager.Factory<PartitionRegionId, ConfigNodeClient>()
+          .createClientManager(new DataNodeClientPoolFactory.ConfigNodeClientPoolFactory());
+
+  public PartitionCache() {
+    this.schemaPartitionCache = Caffeine.newBuilder().maximumSize(cacheSize).build();
+    this.dataPartitionCache = Caffeine.newBuilder().maximumSize(cacheSize).build();
+    this.partitionExecutor =
+        SeriesPartitionExecutor.getSeriesPartitionExecutor(
+            this.seriesSlotExecutorName, this.seriesPartitionSlotNum);
+  }
+
+  // region storage group cache
+
+  /** get storage group of device, return null if not cache */
+  public String getStorageGroup(String devicePath) {
+    synchronized (storageGroupCache) {
```

Comment:
Shall we add the read-write lock here ?

Reply:
this function will change to **private** because there are no need to get just one devicePath outside PartitionCache. Besides, getStorageGroupToDevice and getDeviceToStorageGroup are the only two **public** functions